File: android/app/src/main/java/com/daengdabang/app/OnDeviceTryOnPlugin.java
package com.daengdabang.app;

import android.app.ActivityManager;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.BatteryManager;
import android.os.Build;
import android.os.PowerManager;
import android.util.Base64;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.FloatBuffer;
import java.security.MessageDigest;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import ai.onnxruntime.OnnxTensor;
import ai.onnxruntime.OrtEnvironment;
import ai.onnxruntime.OrtSession;

@CapacitorPlugin(name = "OnDeviceTryOn")
public class OnDeviceTryOnPlugin extends Plugin {
    private static final String MODEL_ASSET = "public/ai/tryon/ddb-lite-tryon-compositor-v2.onnx";
    private static final String MODEL_SHA256 = "3d09e7795872e60f381f02bce0993c2be6ca75fa5530f64ad50979f6315b9eea";
    private static final int EDGE = 256;
    private static final int PLANE = EDGE * EDGE;
    private static final int MAX_DATA_URL_CHARACTERS = 14_000_000;
    private static final int MAX_DECODED_IMAGE_BYTES = 10_000_000;
    private static final int MAX_SOURCE_EDGE = 2048;
    private final ExecutorService inferenceExecutor = Executors.newSingleThreadExecutor();

    @PluginMethod
    public void getCapabilities(PluginCall call) {
        Context context = getContext();
        ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
        ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        activityManager.getMemoryInfo(memoryInfo);
        BatteryManager batteryManager = (BatteryManager) context.getSystemService(Context.BATTERY_SERVICE);
        PowerManager powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        int batteryPercent = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
        boolean charging = batteryManager.isCharging();
        String digest = bundledModelDigest();
        boolean modelAvailable = MODEL_SHA256.equals(digest);
        JSObject result = new JSObject();
        result.put("available", Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1 && modelAvailable);
        result.put("provider", "nnapi");
        result.put("runtimeVersion", OrtEnvironment.getEnvironment().getVersion());
        result.put("modelAvailable", modelAvailable);
        result.put("modelSha256", digest);
        result.put("totalMemoryMb", memoryInfo.totalMem / 1024L / 1024L);
        result.put("batteryLevel", batteryPercent >= 0 ? batteryPercent / 100.0 : -1.0);
        result.put("charging", charging);
        result.put("powerSaveMode", powerManager.isPowerSaveMode());
        result.put("thermalState", thermalState(powerManager));
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O_MR1) result.put("reason", "nnapi_requires_android_8_1");
        else if (!modelAvailable) result.put("reason", "model_integrity_failed");
        call.resolve(result);
    }

    @PluginMethod
    public void run(PluginCall call) {
        String petDataUrl = call.getString("petDataUrl", "");
        String productDataUrl = call.getString("productDataUrl", "");
        String layout = call.getString("layout", "torso");
        String expectedDigest = call.getString("modelSha256", "");
        if (!MODEL_SHA256.equals(expectedDigest) || !MODEL_SHA256.equals(bundledModelDigest())) {
            call.reject("model_integrity_failed", "MODEL_INTEGRITY_FAILED");
            return;
        }
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O_MR1) {
            call.reject("runtime_unavailable", "NNAPI_UNAVAILABLE");
            return;
        }
        inferenceExecutor.execute(() -> {
            long started = System.nanoTime();
            try {
                Bitmap pet = decodeDataUrl(petDataUrl);
                Bitmap product = decodeDataUrl(productDataUrl);
                Prepared prepared = prepare(pet, product, layout);
                float[] output = infer(prepared);
                Bitmap rendered = outputBitmap(output);
                ByteArrayOutputStream stream = new ByteArrayOutputStream();
                rendered.compress(Bitmap.CompressFormat.JPEG, 88, stream);
                JSObject result = new JSObject();
                result.put("imageDataUrl", "data:image/jpeg;base64," + Base64.encodeToString(stream.toByteArray(), Base64.NO_WRAP));
                result.put("provider", "nnapi");
                result.put("runtimeVersion", OrtEnvironment.getEnvironment().getVersion());
                result.put("durationMs", (System.nanoTime() - started) / 1_000_000.0);
                result.put("modelSha256", MODEL_SHA256);
                call.resolve(result);
            } catch (Exception error) {
                call.reject("inference_failed", "ON_DEVICE_INFERENCE_FAILED", error);
            }
        });
    }

    private float[] infer(Prepared prepared) throws Exception {
        byte[] model = readModel();
        OrtEnvironment environment = OrtEnvironment.getEnvironment();
        try (OrtSession.SessionOptions options = new OrtSession.SessionOptions()) {
            options.setIntraOpNumThreads(1);
            options.addNnapi();
            try (OrtSession session = environment.createSession(model, options);
                 OnnxTensor petTensor = OnnxTensor.createTensor(environment, FloatBuffer.wrap(prepared.petRgb), new long[]{1, 3, EDGE, EDGE});
                 OnnxTensor productTensor = OnnxTensor.createTensor(environment, FloatBuffer.wrap(prepared.productRgb), new long[]{1, 3, EDGE, EDGE});
                 OnnxTensor alphaTensor = OnnxTensor.createTensor(environment, FloatBuffer.wrap(prepared.alpha), new long[]{1, 1, EDGE, EDGE})) {
                Map<String, OnnxTensor> inputs = new HashMap<>();
                inputs.put("pet_rgb", petTensor);
                inputs.put("product_rgb", productTensor);
                inputs.put("alpha", alphaTensor);
                try (OrtSession.Result result = session.run(inputs)) {
                    float[][][][] values = (float[][][][]) result.get("result_rgb").orElseThrow().getValue();
                    float[] flat = new float[PLANE * 3];
                    for (int channel = 0; channel < 3; channel++) {
                        for (int y = 0; y < EDGE; y++) {
                            System.arraycopy(values[0][channel][y], 0, flat, channel * PLANE + y * EDGE, EDGE);
                        }
                    }
                    return flat;
                }
            }
        }
    }

    private Prepared prepare(Bitmap pet, Bitmap product, String layout) {
        Bitmap petCanvas = Bitmap.createBitmap(EDGE, EDGE, Bitmap.Config.ARGB_8888);
        Canvas petPainter = new Canvas(petCanvas);
        petPainter.drawColor(Color.rgb(245, 245, 244));
        drawContained(petPainter, pet, new RectF(0, 0, EDGE, EDGE));
        Bitmap productCanvas = Bitmap.createBitmap(EDGE, EDGE, Bitmap.Config.ARGB_8888);
        Canvas productPainter = new Canvas(productCanvas);
        for (RectF rect : layoutRects(layout)) drawContained(productPainter, product, rect);
        int[] petPixels = new int[PLANE];
        int[] productPixels = new int[PLANE];
        petCanvas.getPixels(petPixels, 0, EDGE, 0, 0, EDGE, EDGE);
        productCanvas.getPixels(productPixels, 0, EDGE, 0, 0, EDGE, EDGE);
        float[] petRgb = new float[PLANE * 3];
        float[] productRgb = new float[PLANE * 3];
        float[] alpha = new float[PLANE];
        for (int index = 0; index < PLANE; index++) {
            int petPixel = petPixels[index];
            int productPixel = productPixels[index];
            int red = Color.red(productPixel);
            int green = Color.green(productPixel);
            int blue = Color.blue(productPixel);
            petRgb[index] = Color.red(petPixel) / 255f;
            petRgb[PLANE + index] = Color.green(petPixel) / 255f;
            petRgb[PLANE * 2 + index] = Color.blue(petPixel) / 255f;
            productRgb[index] = red / 255f;
            productRgb[PLANE + index] = green / 255f;
            productRgb[PLANE * 2 + index] = blue / 255f;
            float whiteness = Math.min(red, Math.min(green, blue)) / 255f;
            float whiteSuppression = Math.max(0f, 1f - Math.max(0f, whiteness - 0.9f) * 10f);
            alpha[index] = Color.alpha(productPixel) / 255f * whiteSuppression * 0.82f;
        }
        pet.recycle();
        product.recycle();
        petCanvas.recycle();
        productCanvas.recycle();
        return new Prepared(petRgb, productRgb, alpha);
    }

    private static void drawContained(Canvas canvas, Bitmap bitmap, RectF target) {
        float scale = Math.min(target.width() / bitmap.getWidth(), target.height() / bitmap.getHeight());
        float width = bitmap.getWidth() * scale;
        float height = bitmap.getHeight() * scale;
        RectF centered = new RectF(
                target.centerX() - width / 2f,
                target.centerY() - height / 2f,
                target.centerX() + width / 2f,
                target.centerY() + height / 2f);
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.FILTER_BITMAP_FLAG);
        canvas.drawBitmap(bitmap, null, centered, paint);
    }

    private static RectF[] layoutRects(String layout) {
        if ("head".equals(layout)) return new RectF[]{new RectF(54, 18, 204, 112)};
        if ("neck".equals(layout)) return new RectF[]{new RectF(78, 62, 184, 120)};
        if ("feet".equals(layout)) return new RectF[]{
                new RectF(43, 185, 85, 237), new RectF(87, 188, 129, 240),
                new RectF(135, 188, 177, 240), new RectF(177, 184, 219, 236)};
        if ("leash".equals(layout)) return new RectF[]{new RectF(62, 55, 220, 187)};
        if ("harness".equals(layout)) return new RectF[]{new RectF(56, 63, 210, 183)};
        return new RectF[]{new RectF(48, 55, 214, 187)};
    }

    private static Bitmap outputBitmap(float[] values) {
        int[] pixels = new int[PLANE];
        for (int index = 0; index < PLANE; index++) {
            int red = Math.round(Math.max(0f, Math.min(1f, values[index])) * 255f);
            int green = Math.round(Math.max(0f, Math.min(1f, values[PLANE + index])) * 255f);
            int blue = Math.round(Math.max(0f, Math.min(1f, values[PLANE * 2 + index])) * 255f);
            pixels[index] = Color.rgb(red, green, blue);
        }
        return Bitmap.createBitmap(pixels, EDGE, EDGE, Bitmap.Config.ARGB_8888);
    }

    private static Bitmap decodeDataUrl(String dataUrl) {
        int comma = dataUrl.indexOf(',');
        if (!dataUrl.startsWith("data:image/") || comma < 0 || dataUrl.length() > MAX_DATA_URL_CHARACTERS) {
            throw new IllegalArgumentException("invalid_image_data_url");
        }
        byte[] bytes = Base64.decode(dataUrl.substring(comma + 1), Base64.DEFAULT);
        if (bytes.length > MAX_DECODED_IMAGE_BYTES) throw new IllegalArgumentException("image_too_large");
        BitmapFactory.Options bounds = new BitmapFactory.Options();
        bounds.inJustDecodeBounds = true;
        BitmapFactory.decodeByteArray(bytes, 0, bytes.length, bounds);
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inSampleSize = 1;
        while (Math.max(bounds.outWidth, bounds.outHeight) / options.inSampleSize > MAX_SOURCE_EDGE) {
            options.inSampleSize *= 2;
        }
        Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length, options);
        if (bitmap == null) throw new IllegalArgumentException("invalid_image_data_url");
        return bitmap;
    }

    private byte[] readModel() throws IOException {
        try (InputStream stream = getContext().getAssets().open(MODEL_ASSET);
             ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[8192];
            int read;
            while ((read = stream.read(buffer)) >= 0) output.write(buffer, 0, read);
            return output.toByteArray();
        }
    }

    private String bundledModelDigest() {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256").digest(readModel());
            StringBuilder hex = new StringBuilder();
            for (byte value : digest) hex.append(String.format("%02x", value));
            return hex.toString();
        } catch (Exception ignored) {
            return "";
        }
    }

    private static String thermalState(PowerManager manager) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return "unknown";
        int status = manager.getCurrentThermalStatus();
        if (status >= PowerManager.THERMAL_STATUS_CRITICAL) return "critical";
        if (status >= PowerManager.THERMAL_STATUS_SEVERE) return "serious";
        if (status >= PowerManager.THERMAL_STATUS_MODERATE) return "fair";
        return "nominal";
    }

    @Override
    protected void handleOnDestroy() {
        inferenceExecutor.shutdownNow();
        super.handleOnDestroy();
    }

    private static final class Prepared {
        final float[] petRgb;
        final float[] productRgb;
        final float[] alpha;

        Prepared(float[] petRgb, float[] productRgb, float[] alpha) {
            this.petRgb = petRgb;
            this.productRgb = productRgb;
            this.alpha = alpha;
        }
    }
}

File: ios/App/App/OnDeviceTryOnPlugin.swift
import Capacitor
import CryptoKit
import Foundation
import ImageIO
import OnnxRuntimeBindings
import UIKit

@objc(OnDeviceTryOnPlugin)
public final class OnDeviceTryOnPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OnDeviceTryOnPlugin"
    public let jsName = "OnDeviceTryOn"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getCapabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "run", returnType: CAPPluginReturnPromise),
    ]

    private let modelName = "ddb-lite-tryon-compositor-v2"
    private let modelDigest = "3d09e7795872e60f381f02bce0993c2be6ca75fa5530f64ad50979f6315b9eea"
    private let edge = 256
    private let maxDataUrlCharacters = 14_000_000
    private let maxDecodedImageBytes = 10_000_000
    private let queue = DispatchQueue(label: "com.daengdabang.tryon.coreml", qos: .userInitiated)

    @objc public func getCapabilities(_ call: CAPPluginCall) {
        UIDevice.current.isBatteryMonitoringEnabled = true
        let path = modelPath()
        let digest = path.flatMap(fileDigest) ?? ""
        let battery = UIDevice.current.batteryLevel
        call.resolve([
            "available": ORTIsCoreMLExecutionProviderAvailable() && digest == modelDigest,
            "provider": "coreml",
            "runtimeVersion": String(describing: ORTVersion()),
            "modelAvailable": digest == modelDigest,
            "modelSha256": digest,
            "totalMemoryMb": ProcessInfo.processInfo.physicalMemory / 1024 / 1024,
            "batteryLevel": battery,
            "charging": UIDevice.current.batteryState == .charging || UIDevice.current.batteryState == .full,
            "powerSaveMode": ProcessInfo.processInfo.isLowPowerModeEnabled,
            "thermalState": thermalState(),
            "reason": digest == modelDigest ? "" : "model_integrity_failed",
        ])
    }

    @objc public func run(_ call: CAPPluginCall) {
        guard call.getString("modelSha256") == modelDigest,
              let path = modelPath(), fileDigest(path) == modelDigest else {
            call.reject("model_integrity_failed", "MODEL_INTEGRITY_FAILED")
            return
        }
        guard ORTIsCoreMLExecutionProviderAvailable() else {
            call.reject("runtime_unavailable", "COREML_UNAVAILABLE")
            return
        }
        guard let petDataUrl = call.getString("petDataUrl"),
              let productDataUrl = call.getString("productDataUrl"),
              let pet = decodeDataUrl(petDataUrl),
              let product = decodeDataUrl(productDataUrl) else {
            call.reject("invalid_image_data_url", "INVALID_IMAGE")
            return
        }
        let layout = call.getString("layout") ?? "torso"
        queue.async { [weak self] in
            guard let self else { return }
            let started = Date()
            do {
                let prepared = try self.prepare(pet: pet, product: product, layout: layout)
                let output = try self.infer(path: path, prepared: prepared)
                let image = try self.outputImage(output)
                guard let jpeg = image.jpegData(compressionQuality: 0.88) else {
                    throw TryOnError.render
                }
                call.resolve([
                    "imageDataUrl": "data:image/jpeg;base64,\(jpeg.base64EncodedString())",
                    "provider": "coreml",
                    "runtimeVersion": String(describing: ORTVersion()),
                    "durationMs": Date().timeIntervalSince(started) * 1000,
                    "modelSha256": self.modelDigest,
                ])
            } catch {
                call.reject("inference_failed", "ON_DEVICE_INFERENCE_FAILED", error)
            }
        }
    }

    private func infer(path: String, prepared: Prepared) throws -> [Float32] {
        let environment = try ORTEnv(loggingLevel: .warning)
        let options = try ORTSessionOptions()
        try options.setIntraOpNumThreads(1)
        let coreML = ORTCoreMLExecutionProviderOptions()
        coreML.enableOnSubgraphs = true
        try options.appendCoreMLExecutionProvider(with: coreML)
        let session = try ORTSession(env: environment, modelPath: path, sessionOptions: options)
        let rgbShape: [NSNumber] = [1, 3, NSNumber(value: edge), NSNumber(value: edge)]
        let alphaShape: [NSNumber] = [1, 1, NSNumber(value: edge), NSNumber(value: edge)]
        let inputs = [
            "pet_rgb": try tensor(prepared.petRGB, shape: rgbShape),
            "product_rgb": try tensor(prepared.productRGB, shape: rgbShape),
            "alpha": try tensor(prepared.alpha, shape: alphaShape),
        ]
        let outputs = try session.run(withInputs: inputs, outputNames: ["result_rgb"], runOptions: nil)
        guard let value = outputs["result_rgb"] else { throw TryOnError.inference }
        let data = try value.tensorData() as Data
        return data.withUnsafeBytes { Array($0.bindMemory(to: Float32.self)) }
    }

    private func tensor(_ values: [Float32], shape: [NSNumber]) throws -> ORTValue {
        let data = values.withUnsafeBufferPointer { Data(buffer: $0) }
        return try ORTValue(
            tensorData: NSMutableData(data: data),
            elementType: .float,
            shape: shape)
    }

    private func prepare(pet: UIImage, product: UIImage, layout: String) throws -> Prepared {
        let petImage = render(size: CGSize(width: edge, height: edge), opaque: true) { context in
            UIColor(red: 245/255, green: 245/255, blue: 244/255, alpha: 1).setFill()
            context.fill(CGRect(x: 0, y: 0, width: edge, height: edge))
            drawContained(pet, in: CGRect(x: 0, y: 0, width: edge, height: edge))
        }
        let productImage = render(size: CGSize(width: edge, height: edge), opaque: false) { _ in
            for rect in layoutRects(layout) { drawContained(product, in: rect) }
        }
        let petPixels = try rgba(petImage)
        let productPixels = try rgba(productImage)
        let plane = edge * edge
        var petRGB = [Float32](repeating: 0, count: plane * 3)
        var productRGB = [Float32](repeating: 0, count: plane * 3)
        var alpha = [Float32](repeating: 0, count: plane)
        for index in 0..<plane {
            let pixel = index * 4
            for channel in 0..<3 {
                petRGB[channel * plane + index] = Float32(petPixels[pixel + channel]) / 255
                productRGB[channel * plane + index] = Float32(productPixels[pixel + channel]) / 255
            }
            let whiteness = Float32(min(productPixels[pixel], min(productPixels[pixel + 1], productPixels[pixel + 2]))) / 255
            let whiteSuppression = max(0, 1 - max(0, whiteness - 0.9) * 10)
            alpha[index] = Float32(productPixels[pixel + 3]) / 255 * whiteSuppression * 0.82
        }
        return Prepared(petRGB: petRGB, productRGB: productRGB, alpha: alpha)
    }

    private func outputImage(_ values: [Float32]) throws -> UIImage {
        let plane = edge * edge
        guard values.count == plane * 3 else { throw TryOnError.render }
        var pixels = [UInt8](repeating: 255, count: plane * 4)
        for index in 0..<plane {
            pixels[index * 4] = byte(values[index])
            pixels[index * 4 + 1] = byte(values[plane + index])
            pixels[index * 4 + 2] = byte(values[plane * 2 + index])
        }
        let data = Data(pixels) as CFData
        guard let provider = CGDataProvider(data: data),
              let cgImage = CGImage(
                width: edge, height: edge, bitsPerComponent: 8, bitsPerPixel: 32,
                bytesPerRow: edge * 4, space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.last.rawValue),
                provider: provider, decode: nil, shouldInterpolate: true, intent: .defaultIntent) else {
            throw TryOnError.render
        }
        return UIImage(cgImage: cgImage)
    }

    private func rgba(_ image: UIImage) throws -> [UInt8] {
        guard let cgImage = image.cgImage else { throw TryOnError.render }
        var bytes = [UInt8](repeating: 0, count: edge * edge * 4)
        guard let context = CGContext(
            data: &bytes, width: edge, height: edge, bitsPerComponent: 8,
            bytesPerRow: edge * 4, space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue) else { throw TryOnError.render }
        context.translateBy(x: 0, y: CGFloat(edge))
        context.scaleBy(x: 1, y: -1)
        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: edge, height: edge))
        return bytes
    }

    private func render(size: CGSize, opaque: Bool, actions: (CGContext) -> Void) -> UIImage {
        let format = UIGraphicsImageRendererFormat()
        format.opaque = opaque
        format.scale = 1
        return UIGraphicsImageRenderer(size: size, format: format).image { renderer in actions(renderer.cgContext) }
    }

    private func drawContained(_ image: UIImage, in target: CGRect) {
        let scale = min(target.width / image.size.width, target.height / image.size.height)
        let size = CGSize(width: image.size.width * scale, height: image.size.height * scale)
        image.draw(in: CGRect(x: target.midX - size.width / 2, y: target.midY - size.height / 2, width: size.width, height: size.height))
    }

    private func layoutRects(_ layout: String) -> [CGRect] {
        switch layout {
        case "head": return [CGRect(x: 54, y: 18, width: 150, height: 94)]
        case "neck": return [CGRect(x: 78, y: 62, width: 106, height: 58)]
        case "feet": return [CGRect(x: 43, y: 185, width: 42, height: 52), CGRect(x: 87, y: 188, width: 42, height: 52), CGRect(x: 135, y: 188, width: 42, height: 52), CGRect(x: 177, y: 184, width: 42, height: 52)]
        case "leash": return [CGRect(x: 62, y: 55, width: 158, height: 132)]
        case "harness": return [CGRect(x: 56, y: 63, width: 154, height: 120)]
        default: return [CGRect(x: 48, y: 55, width: 166, height: 132)]
        }
    }

    private func decodeDataUrl(_ value: String) -> UIImage? {
        guard value.count <= maxDataUrlCharacters,
              value.hasPrefix("data:image/"), let comma = value.firstIndex(of: ","),
              let data = Data(base64Encoded: String(value[value.index(after: comma)...])),
              data.count <= maxDecodedImageBytes,
              let source = CGImageSourceCreateWithData(data as CFData, nil),
              let image = CGImageSourceCreateThumbnailAtIndex(source, 0, [
                kCGImageSourceCreateThumbnailFromImageAlways: true,
                kCGImageSourceCreateThumbnailWithTransform: true,
                kCGImageSourceThumbnailMaxPixelSize: 2048,
              ] as CFDictionary) else { return nil }
        return UIImage(cgImage: image)
    }

    private func modelPath() -> String? {
        Bundle.main.path(forResource: modelName, ofType: "onnx", inDirectory: "public/ai/tryon")
    }

    private func fileDigest(_ path: String) -> String? {
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: path), options: .mappedIfSafe) else { return nil }
        return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    private func byte(_ value: Float32) -> UInt8 { UInt8((max(0, min(1, value)) * 255).rounded()) }

    private func thermalState() -> String {
        switch ProcessInfo.processInfo.thermalState {
        case .nominal: return "nominal"
        case .fair: return "fair"
        case .serious: return "serious"
        case .critical: return "critical"
        @unknown default: return "unknown"
        }
    }
}

private struct Prepared {
    let petRGB: [Float32]
    let productRGB: [Float32]
    let alpha: [Float32]
}

private enum TryOnError: Error { case inference, render }
